### ISSUE TITLE
MultiPolygon interprets second polygon as hole

### ISSUE THREAD

**Issue by ChrisWGeorge** [qgis/QGIS#54511]
### What is the bug or the crash?

I have an example of a multipolygon with two parts.  The parts are non-overlapping, though they do share one vertex.  (I don't think the common vertex is significant.)  The polygon is written to a shapefile using QgsFeature.setGeometry.  When the shapefile is read there is an error 'Hole lies outside shell' and only one part is read.
[multipolygon_error.txt](https://github.com/qgis/QGIS/files/12537174/multipolygon_error.txt)

### Steps to reproduce the issue

Read the attached shapefile and run Check Validity
[dem_liljwshed.zip](https://github.com/qgis/QGIS/files/12537286/dem_liljwshed.zip)

The error is in writing the shapefile, not reading it.  In QGIS 3.22.12 there is no problem.  Running the same code on the same project in qgis.3.28.10 gives the error, and the shapefile written by 3.28 reports the same errors when read by 3.22.

The text file above is the geometry of the shape with PolygonId 149.

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

QGIS version | 3.28.10-Firenze | QGIS code revision | e2edb9c468
-- | -- | -- | --
Qt version | 5.15.3
Python version | 3.9.5
GDAL/OGR version | 3.7.1
PROJ version | 9.2.1
EPSG Registry database version | v10.088 (2023-05-13)
GEOS version | 3.12.0-CAPI-1.18.0
SQLite version | 3.41.1
PDAL version | 2.5.5
PostgreSQL client version | unknown
SpatiaLite version | 5.1.0
QWT version | 6.1.6
QScintilla2 version | 2.13.1
OS version | Windows 10 Version 2009
  |   |   |  
Active Python plugins
plugin_reloader | 0.9.3
QAPEX | 1.0.4
QSWAT3_9 | 1.6.5
QSWATPlus3_9 | 2.4.5
remotedebug | 0.4.3
grassprovider | 2.12.99
processing | 2.12.99
sagaprovider | 2.12.99

</body></html>QGIS version
3.28.10-Firenze
QGIS code revision
[e2edb9c468](https://github.com/qgis/QGIS/commit/e2edb9c468)
Qt version
5.15.3
Python version
3.9.5
GDAL/OGR version
3.7.1
PROJ version
9.2.1
EPSG Registry database version
v10.088 (2023-05-13)
GEOS version
3.12.0-CAPI-1.18.0
SQLite version
3.41.1
PDAL version
2.5.5
PostgreSQL client version
unknown
SpatiaLite version
5.1.0
QWT version
6.1.6
QScintilla2 version
2.13.1
OS version
Windows 10 Version 2009

Active Python plugins
plugin_reloader
0.9.3
QAPEX
1.0.4
QSWAT3_9
1.6.5
QSWATPlus3_9
2.4.5
remotedebug
0.4.3
grassprovider
2.12.99
processing
2.12.99
sagaprovider
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

**Comment by agiudiceandrea** [comment]
@ChrisWGeorge, thanks for reporting. Could you please share a minimal version of your script in order to replicate the issue?

**Comment by ChrisWGeorge** [comment]
This script writes a file 'polybug.shp' to the user's Documents folder (WARNING overwriting any existing such file).  It contains one shape consisting of two disjoint unit squares.  It checks geometry validity before writing, reporting 'No creation errors'.  It then reads the file and checks the geometry.  In QGIS 3.28.11 it reports an error, in QGIS 3.22.12 it reports 'No read errors'.
[polybug.zip](https://github.com/qgis/QGIS/files/12930459/polybug.zip)

**Comment by ChrisWGeorge** [comment]
Sorry - did not mean to close.  Just submitted a script to illustarte the bug.

**Comment by agiudiceandrea** [comment]
@ChrisWGeorge, thanks.

I've tested the script using the latest QGIS 3.33.0-Master and it creates an ESRI Shapefile layer with the correct geometry and it outputs:
```
No creation errors
No read errors
```
The issue has been fixed in QGIS with PR https://github.com/qgis/QGIS/pull/54761 and in the GDAL/OGR library with PR https://github.com/OSGeo/gdal/pull/8483. The fix is available in current QGIS 3.33.0-Master and will be available in the upcoming QGIS 3.28.12 and QGIS 3.34.0 versions.
